src/query_builder.rs

```diff
@@ -0,0 +1,384 @@
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![macro_use]
+use std::collections::{
+    BTreeMap,
+};
+
+use mentat_core::{
+    Entid,
+    NamespacedKeyword,
+    TypedValue,
+    ValueType,
+};
+
+use ::{
+    HasSchema,
+    Queryable,
+    QueryInputs,
+    QueryOutput,
+    Store,
+    Variable,
+};
+
+use errors::{
+    ErrorKind,
+    Result,
+};
+
+pub struct QueryBuilder<'a> {
+    sql: String,
+    values: BTreeMap<Variable, TypedValue>,
+    types: BTreeMap<Variable, ValueType>,
+    store: &'a mut Store,
+}
+
+impl<'a> QueryBuilder<'a> {
+    pub fn new<T>(store: &'a mut Store, sql: T) -> QueryBuilder where T: Into<String> {
+        QueryBuilder { sql: sql.into(), values: BTreeMap::new(), types: BTreeMap::new(), store }
+    }
+
+    pub fn bind_value<T>(mut self, var: &str, value: T) -> Self where T: Into<TypedValue> {
+        self.values.insert(Variable::from_valid_name(var), value.into());
+        self
+    }
+
+    pub fn bind_ref_from_kw(mut self, var: &str, value: NamespacedKeyword) -> Result<Self> {
+        let entid = self.store.conn().current_schema().get_entid(&value).ok_or(ErrorKind::UnknownAttribute(value.to_string()))?;
+        self.values.insert(Variable::from_valid_name(var), TypedValue::Ref(entid.into()));
+        Ok(self)
+    }
+
+    pub fn bind_ref<T>(mut self, var: &str, value: T) -> Self where T: Into<Entid> {
+       self.values.insert(Variable::from_valid_name(var), TypedValue::Ref(value.into()));
+       self
+    }
+
+    pub fn bind_instant(mut self, var: &str, value: i64) -> Self {
+       self.values.insert(Variable::from_valid_name(var), TypedValue::instant(value));
+       self
+    }
+
+    pub fn bind_type(mut self, var: &str, value_type: ValueType) -> Self {
+        self.types.insert(Variable::from_valid_name(var), value_type);
+        self
+    }
+
+    pub fn execute(&mut self) -> Result<QueryOutput> {
+        let values = ::std::mem::replace(&mut self.values, Default::default());
+        let types = ::std::mem::replace(&mut self.types, Default::default());
+        let query_inputs = QueryInputs::new(types, values)?;
+        let read = self.store.begin_read()?;
+        read.q_once(&self.sql, query_inputs)
+    }
+
+    pub fn execute_scalar(mut self) -> Result<Option<TypedValue>> {
+        let results = self.execute()?;
+        results.into_scalar().map_err(|e| e.into())
+    }
+
+    pub fn execute_coll(mut self) -> Result<Vec<TypedValue>> {
+        let results = self.execute()?;
+        results.into_coll().map_err(|e| e.into())
```

Comment:
I think you probably want to import the `IntoResult` trait and use `into_coll_result` and friends.